### PR TITLE
Add search endpoint for document chunks by text similarity

### PR DESCRIPTION
**Pull request by g-fiorina**
- Implemented a new /search GET endpoint in api.py that allows users to search for document chunks similar to a given text.
- Added get_similar_chunks function to handle the similarity search using OpenAI embeddings.
- Included error handling for invalid inputs and unexpected errors.

### CODE REVIEW

File: ingest_documents_api/api.py [hunk 1]
```diff
@@ -86,3 +100,25 @@ def get_document(document_id: int):
             status_code=500,
             detail="An unexpected error occurred while retrieving the document",
         ) from e
+
+        
+@app.get("/search", response_model=List[RetrievedChunk])
+def search_by_similarity(
+    text: str, minimum_similarity: float = -0.8, max_quantity_chunks: int = None
```
Comment (fede-bello) [inline ingest_documents_api/api.py:107]: What do you think about making this a constant to improve readability? 
Also, I think the value might be a little bit to high, I would test this a bit more

File: ingest_documents_api/database.py [hunk 2]
```diff
@@ -105,3 +104,33 @@ def save_document(pdf_link: str):
...
+):
+
+    embeddings = OpenAIEmbeddings(model="text-embedding-3-small")
+    embedded_text = embeddings.embed_query(text)
+
+    query = (
+        DocumentChunk.select(
+            DocumentChunk.id,
+            DocumentChunk.chunk_number,
+            DocumentChunk.document_id.alias("document_id"),
```
Comment (fede-bello) [inline ingest_documents_api/database.py:120]: Why is the alias needed?

File: ingest_documents_api/database.py [hunk 3]
```diff
@@ -105,3 +104,33 @@ def save_document(pdf_link: str):
...
+            DocumentChunk.text,
+            DocumentChunk.embedding.max_inner_product(embedded_text).alias(
```
Comment (fede-bello) [inline ingest_documents_api/database.py:122]: What do you thinkg about defining a `similarity = DocumentChunk.embedding.max_inner_product(embedded_text) ` and using that in the select, where and order by?

File: ingest_documents_api/database.py [hunk 4]
```diff
@@ -105,3 +104,33 @@ def save_document(pdf_link: str):
...
+            ),
+        )
+        .where(
+            DocumentChunk.embedding.max_inner_product(embedded_text)
+            < minimum_similarity
+        )
+        .order_by(DocumentChunk.embedding.max_inner_product(embedded_text))
+    )
+
+    if max_quantity_chunks:
```
Comment (fede-bello) [inline ingest_documents_api/database.py:133]: I think its better to always define a limit and if the parameter is sent to override that limite.
Without a limit you ran the risk of running out of memory, taking too long in the query or eventually passing way to much context to the LLM

**Review by fede-bello — CHANGES_REQUESTED** [review]
I really liked the approach Guille, I left some comments

File: ingest_documents_api/database.py (hunk 2 above)
Comment (g-fiorina) [reply ingest_documents_api/database.py:120]: It is a mistake, I left it while doing some changes and never removed it. I will erase it, thank you!

File: ingest_documents_api/database.py (hunk 3 above)
Comment (g-fiorina) [reply ingest_documents_api/database.py:122]: Amazing, much clearer!

**Review by fede-bello — APPROVED** [review]
Nice

**Pull request merged**